File: app/src/test/java/com/kickstarter/ui/intents/ProjectIntentMapperTest.java
package com.kickstarter.ui.intents;

import android.content.Intent;
import android.net.Uri;

import com.kickstarter.KSRobolectricTestCase;
import com.kickstarter.mock.factories.ProjectFactory;
import com.kickstarter.mock.factories.PushNotificationEnvelopeFactory;
import com.kickstarter.libs.RefTag;
import com.kickstarter.models.Project;
import com.kickstarter.mock.services.MockApiClient;
import com.kickstarter.services.apiresponses.PushNotificationEnvelope;
import com.kickstarter.ui.IntentKey;
import com.kickstarter.ui.intentmappers.ProjectIntentMapper;

import org.junit.Test;

import rx.observers.TestSubscriber;

public final class ProjectIntentMapperTest extends KSRobolectricTestCase {
  @Test
  public void testProject_emitsFromProjectParamExtra() {
    final Intent intent = new Intent().putExtra(IntentKey.PROJECT_PARAM, "skull-graphic-tee");

    final TestSubscriber<Project> resultTest = TestSubscriber.create();
    ProjectIntentMapper.project(intent, new MockApiClient())
      .subscribe(resultTest);

    resultTest.assertValueCount(1);
  }

  @Test
  public void testProject_emitsTwiceFromProjectExtra() {
    final Project project = ProjectFactory.project();
    final Intent intent = new Intent().putExtra(IntentKey.PROJECT, project);

    final TestSubscriber<Project> resultTest = TestSubscriber.create();
    ProjectIntentMapper.project(intent, new MockApiClient())
      .subscribe(resultTest);

    resultTest.assertValueCount(2);
  }

  @Test
  public void testProject_emitsFromKsrProjectUri() {
    final Uri uri = Uri.parse("ksr://www.kickstarter.com/projects/1186238668/skull-graphic-tee");
    final Intent intent = new Intent(Intent.ACTION_VIEW, uri);

    final TestSubscriber<Project> resultTest = TestSubscriber.create();
    ProjectIntentMapper.project(intent, new MockApiClient())
      .subscribe(resultTest);

    resultTest.assertValueCount(1);
  }

  @Test
  public void testProject_emitsFromHttpsProjectUri() {
    final Uri uri = Uri.parse("https://www.kickstarter.com/projects/1186238668/skull-graphic-tee");
    final Intent intent = new Intent(Intent.ACTION_VIEW, uri);

    final TestSubscriber<Project> resultTest = TestSubscriber.create();
    ProjectIntentMapper.project(intent, new MockApiClient())
      .subscribe(resultTest);

    resultTest.assertValueCount(1);
  }

  @Test
  public void testRefTag_emitsFromRefTag() {
    final RefTag refTag = RefTag.from("test");
    final Intent intent = new Intent().putExtra(IntentKey.REF_TAG, refTag);

    final TestSubscriber<RefTag> resultTest = TestSubscriber.create();
    ProjectIntentMapper.refTag(intent).subscribe(resultTest);

    resultTest.assertValue(refTag);
  }

  @Test
  public void testRefTag_emitsNullWithNoRefTag() {
    final Intent intent = new Intent();

    final TestSubscriber<RefTag> resultTest = TestSubscriber.create();
    ProjectIntentMapper.refTag(intent).subscribe(resultTest);

    resultTest.assertValue(null);
  }

  @Test
  public void testPushNotificationEnvelope_emitsFromEnvelope() {
    final PushNotificationEnvelope envelope = PushNotificationEnvelopeFactory.envelope();
    final Intent intent = new Intent().putExtra(IntentKey.PUSH_NOTIFICATION_ENVELOPE, envelope);

    final TestSubscriber<PushNotificationEnvelope> resultTest = TestSubscriber.create();
    ProjectIntentMapper.pushNotificationEnvelope(intent).subscribe(resultTest);

    resultTest.assertValue(envelope);
  }

  @Test
  public void testPushNotificationEnvelope_doesNotEmitWithoutEnvelope() {
    final PushNotificationEnvelope envelope = PushNotificationEnvelopeFactory.envelope();
    final Intent intent = new Intent();

    final TestSubscriber<PushNotificationEnvelope> resultTest = TestSubscriber.create();
    ProjectIntentMapper.pushNotificationEnvelope(intent).subscribe(resultTest);

    resultTest.assertNoValues();
  }
}

File: app/src/test/java/com/kickstarter/ui/intents/ProjectIntentMapperTest.kt
package com.kickstarter.ui.intents

import android.content.Intent
import android.net.Uri
import androidx.core.net.toUri
import com.kickstarter.KSRobolectricTestCase
import com.kickstarter.libs.RefTag
import com.kickstarter.libs.utils.KsOptional
import com.kickstarter.libs.utils.extensions.addToDisposable
import com.kickstarter.mock.factories.ProjectFactory
import com.kickstarter.mock.services.MockApiClientV2
import com.kickstarter.mock.services.MockApolloClientV2
import com.kickstarter.models.Project
import com.kickstarter.ui.IntentKey
import com.kickstarter.ui.intentmappers.ProjectIntentMapper
import io.reactivex.disposables.CompositeDisposable
import org.junit.After
import org.junit.Test

class ProjectIntentMapperTest : KSRobolectricTestCase() {
    private val disposables = CompositeDisposable()

    @Test
    fun testProject_emitsFromProjectParamExtra() {
        val intent = Intent().putExtra(IntentKey.PROJECT_PARAM, "skull-graphic-tee")
        val resultTestV2 = io.reactivex.subscribers.TestSubscriber.create<Project>()

        attacheTestResultSubscriber(intent, resultTestV2)

        resultTestV2.assertValueCount(1)
    }

    @Test
    fun testProject_emitsFromProjectParamExtraApollo() {
        val intent = Intent().putExtra(IntentKey.PROJECT_PARAM, "skull-graphic-tee")
        val resultTestV2 = io.reactivex.subscribers.TestSubscriber.create<Project>()

        attacheTestResultSubscriber(intent, resultTestV2)

        resultTestV2.assertValueCount(1)
    }

    @Test
    fun testProject_emitsTwiceFromProjectExtra_V2() {
        val project = ProjectFactory.project()
        val intent = Intent().putExtra(IntentKey.PROJECT, project)
        val resultTest = io.reactivex.subscribers.TestSubscriber.create<Project>()
        ProjectIntentMapper.project(intent, MockApiClientV2())
            .subscribe { resultTest.onNext(it) }.addToDisposable(disposables)
        resultTest.assertValueCount(2)
    }

    @Test
    fun testProject_emitsFromKsrProjectUri() {
        val uri = Uri.parse("ksr://www.kickstarter.com/projects/1186238668/skull-graphic-tee")
        val intent = Intent(Intent.ACTION_VIEW, uri)
        val resultTestV2 = io.reactivex.subscribers.TestSubscriber.create<Project>()

        attacheTestResultSubscriber(intent, resultTestV2)

        resultTestV2.assertValueCount(1)
    }

    @Test
    fun testProject_emitsFromHttpsProjectUri_V2() {
        val uri = Uri.parse("https://www.kickstarter.com/projects/1186238668/skull-graphic-tee")
        val intent = Intent(Intent.ACTION_VIEW, uri)
        val resultTest = io.reactivex.subscribers.TestSubscriber.create<Project>()
        ProjectIntentMapper.project(intent, MockApiClientV2())
            .subscribe { resultTest.onNext(it) }.addToDisposable(disposables)
        resultTest.assertValueCount(1)
    }

    @Test
    fun testRefTag_emitsFromRefTag() {
        val refTag = RefTag.from("test")
        val intent = Intent().putExtra(IntentKey.REF_TAG, refTag)
        val resultTest = io.reactivex.subscribers.TestSubscriber.create<KsOptional<RefTag?>>()
        ProjectIntentMapper.refTag(intent).subscribe { resultTest.onNext(it) }.addToDisposable(disposables)
        assertEquals(resultTest.values().get(0).getValue(), refTag)
    }

    @Test
    fun testRefTag_emitsNullWithNoRefTag() {
        val intent = Intent()
        val resultTest = io.reactivex.subscribers.TestSubscriber.create<KsOptional<RefTag?>>()
        ProjectIntentMapper.refTag(intent).subscribe { resultTest.onNext(it) }.addToDisposable(disposables)
        resultTest.assertValueCount(1)
        assertNull(resultTest.values().get(0).getValue())
    }

    @Test
    fun testProjectHasSaveQueryFromUri() {
        assertTrue(ProjectIntentMapper.hasSaveQueryFromUri("ksr://www.kickstarter.com/projects/1186238668/skull-graphic-tee?save=true".toUri()))
        assertFalse(ProjectIntentMapper.hasSaveQueryFromUri("ksr://www.kickstarter.com/projects/1186238668/skull-graphic-tee".toUri()))
    }

    @After
    fun cleanUp() {
        disposables.clear()
    }
    private fun attacheTestResultSubscriber(
        intent: Intent,
        resultTestV2: io.reactivex.subscribers.TestSubscriber<Project>
    ) {

        ProjectIntentMapper.project(intent, MockApolloClientV2()).subscribe {
            resultTestV2.onNext(it)
        }.addToDisposable(disposables)
    }
}
